src/run/run.c

```diff
@@ -1720,11 +1720,14 @@ static int run(int argc, char* argv[]) {
         if (r <= 0)
                 return r;
 
-        if (!strv_isempty(arg_cmdline) && arg_transport == BUS_TRANSPORT_LOCAL) {
-                _cleanup_free_ char *command = NULL;
-
-                /* Patch in an absolute path */
+        if (!strv_isempty(arg_cmdline) &&
+            arg_transport == BUS_TRANSPORT_LOCAL &&
+            !strv_find_startswith(arg_property, "RootDirectory=") &&
+            !strv_find_startswith(arg_property, "RootImage=")) {
+                /* Patch in an absolute path to fail early for user convenience, but only when
+                 * we can do it (i.e. we will be running from the same file system). */
 
+                _cleanup_free_ char *command = NULL;
                 r = find_executable(arg_cmdline[0], &command);
```

Comment:
why do this client-side? we used to do that for convenience, so that people can specify relative names, but given that PID 1 now resolves relative names, we can drop the whole find_executable() call here altogether, and just let pid 1 do it?

Reply:
The user might have $PATH set, and then we want to resolve the thing locally using their $PATH. Also, it's nicer to fail immediately without going over the bus if there's a typo in the executable name and such. And using systemd-run without RootImage=/RootDirectory/-M/-H is probably the most common case, 99% of invocations, so it makes sense to optimize for it.

Reply:
Also, we can resolve a relative *path* (not a single name component) here.

Reply:
but it's so magic... you never know what you end up with. maybe add a runtime switch --resolve-executable=client|manager|auto` where people can configure how things are resolved, and we'd default to `auto` which would work client client except if we see RootImage=/RootDirectory=, ... then at least it wouldn't be so magic, but explicit on request.

Reply:
hmm, but afaics if i use RootImage= with this version, do things even work then? i.e. PID 1 refuses relative paths specified via the dbus iface, no?

Reply:
Hmm, I tested the reproducer specified in #13338. (BTW, why doesn't github "link" the PR to that issue. It's smart enough to provide a backlink in the issue, but not to mark the patch as resolving the issue. Strange.)